.github/workflows/tests.yml

```diff
@@ -67,22 +87,316 @@ jobs:
+        path: deps
+        key: edbdeps-pip-${{ hashFiles('setup.py') }}
+
+    - name: Install Python deps
+      run: |
+        pip install -U --no-index --find-links=deps deps/*
+
+    - name: Download cache key
+      uses: actions/download-artifact@v2
+      with:
+        name: build-cache-key-and-time-stats
+        path: .tmp
+
+    - name: Handle build cache
+      uses: actions/cache@v2
+      id: buildcache
+      with:
+        path: build
+        key: edbbuild-v2-${{ hashFiles('.tmp/build_cache_key.txt') }}
+
+    - name: Stop if we cannot retrieve the build cache
+      if: steps.buildcache.outputs.cache-hit != 'true'
+      run: |
+        echo ::error::Cannot retrieve build cache.
+        exit 1
+
+    - name: Install rust toolchain
+      uses: actions-rs/toolchain@v1
+      with:
+        profile: minimal
+        toolchain: stable
+        default: true
+
     - name: Cargo test
       uses: actions-rs/cargo@v1
       with:
         command: test
+
+  python-test:
+    needs: build
+    runs-on: ubuntu-18.04
+    strategy:
+      matrix:
+        shard: [
+             1/32,  2/32,  3/32,  4/32,
+             5/32,  6/32,  7/32,  8/32,
+             9/32, 10/32, 11/32, 12/32,
+            13/32, 14/32, 15/32, 16/32,
+            17/32, 18/32, 19/32, 20/32,
+            21/32, 22/32, 23/32, 24/32,
+            25/32, 26/32, 27/32, 28/32,
+            29/32, 30/32, 31/32, 32/32,
+        ]
+
+    steps:
+    - uses: actions/checkout@v2
+      with:
+        fetch-depth: 0
+        submodules: false
+
+    - uses: actions/checkout@v2
+      with:
+        fetch-depth: 50
+        submodules: true
+
+    - name: Set up Python
+      uses: actions/setup-python@v2
+      with:
+        python-version: 3.8
+
+    - name: Handle Pip dependency cache
+      uses: actions/cache@v2
+      id: depcache
+      with:
+        path: deps
+        key: edbdeps-pip-${{ hashFiles('setup.py') }}
+
+    - name: Install Python deps
+      run: |
+        pip install -U --no-index --find-links=deps deps/*
+
+    - name: Download shared artifacts
+      uses: actions/download-artifact@v2
+      with:
+        name: build-cache-key-and-time-stats
+        path: .tmp
+
+    - name: Handle build cache
+      uses: actions/cache@v2
+      id: buildcache
+      with:
+        path: build
+        key: edbbuild-v2-${{ hashFiles('.tmp/build_cache_key.txt') }}
+
+    - name: Stop if we cannot retrieve the build cache
+      if: steps.buildcache.outputs.cache-hit != 'true'
+      run: |
+        echo ::error::Cannot retrieve build cache.
+        exit 1
+
+    - name: Install rust toolchain
+      uses: actions-rs/toolchain@v1
+      with:
+        profile: minimal
+        toolchain: stable
+        default: true
+
+    - name: Install
+      run: |
+        # --no-use-pep517 because we have explicitly installed all deps
+        # and don't want them to be reinstalled in an "isolated env".
+        pip install --no-use-pep517 --no-deps -e .[test,docs]
```

Comment:
I wonder if we could cache the entire virtualenv too to avoid doing this step in shards?